### ISSUE TITLE
linter: unbound-method false positives on non-class object destructuring

### ISSUE THREAD

**Issue by viluMatilda** [oxc-project/oxc#21496]
## Summary

`typescript/unbound-method` flags destructured properties from plain objects, hook returns, and function parameters as unbound methods, even though they are not class instance methods and have no `this` binding concern.

This produces false positives that are common in React/TypeScript codebases, where destructuring from hooks and plain objects is idiomatic.

## Examples of false positives

All examples below are from a real codebase where we attempted to enable this rule.

### 1. Destructuring from a custom hook return

```typescript
// hooks/useUpdateSearchParams.ts returns a plain object with functions
// ❌ Flagged: "Avoid referencing unbound methods which may cause unintentional scoping of this"
const { replaceSearchParam } = useUpdateSearchParams();
```

```typescript
// useFormFieldWarning returns { warning, triggerWarning }
// ❌ Flagged on triggerWarning
const { warning, triggerWarning } = useFormFieldWarning<StockTakingFormData, \`items.\${number}.amount\`>({
  name: \`items.\${index}.amount\`,
  validate: (value) => /* ... */,
});
```

### 2. Destructuring a function parameter's property with rename

```typescript
// The function parameter is a plain object type, not a class instance.
// ❌ Flagged: the renamed `validate` property triggers the rule
export function useFormFieldWarning<
  TFieldValues extends Record<string, unknown>,
  TFieldName extends FieldPath<TFieldValues>,
>({
  name,
  validate: validateValue, // ← flagged here
}: {
  name: TFieldName;
  validate(value: FieldPathValue<TFieldValues, TFieldName>): string | null | undefined;
}): {
  warning: string | undefined;
  triggerWarning(): void;
} {
  // validateValue is used as a plain function — no `this` involved
  const validateValueRef = useRef(validateValue);
  // ...
}
```

### 3. Property access on a plain object in tests

```typescript
// apiClient is a plain object literal with method properties, not a class instance
// ❌ Flagged: "This reference may be unbound and lose this context"
vi.mocked(apiClient.put).mockResolvedValue({});
```

## Proposal

Add a configuration option (e.g. `ignoreNonClassDestructuring` or `ignoreObjectDestructuring`) that suppresses the warning when the destructured source is not a class instance.

The rule would still flag:
- `instance.method` references from class instances
- Destructured methods from class instances (`const { method } = classInstance`)
- Callbacks like `[1,2,3].map(instance.method)`

But it would **not** flag:
- Destructured properties from plain objects, hook returns, or function parameters
- Property access on non-class objects (e.g. `apiClient.put` where `apiClient` is a plain object literal)

This would allow the classic typescript-eslint behavior to apply for real \`this\`-binding risks while reducing noise in modern React/TypeScript codebases where destructuring from hooks and plain objects is the norm.

## Context

The upstream `typescript-eslint` rule has this same limitation (see [typescript-eslint#8947](https://github.com/typescript-eslint/typescript-eslint/issues/8947)), but since oxlint has type-aware linting, it could potentially distinguish between class instances and plain objects to offer a more accurate rule.

**Comment by camc314** [comment]
Hi - thanks for making this issue.

> unbound methods, even though they are not class instance methods and have no this binding concern.

This doesn't make sense. Objects do have a `this` binding concern. Accessing the `this` property within an object method returns the object, if you destructure, you change the `this` binding which is exactly what this rule is trying to prevent. to my knowledge, typescript has no knowledge of if a `this` binding is used within an object/class method, so it's impossible for us to statically tell if this is the case.

> The upstream typescript-eslint rule has this same limitation (see https://github.com/typescript-eslint/typescript-eslint/issues/8947), but since oxlint has type-aware linting, it could potentially distinguish between class instances and plain objects to offer a more accurate rule.

This doesn't make sense. typescript-eslint uses type aware linting, and it's impossible to differentiate between class vs plain objects.

As per the upstream typescript-eslint issue, this is the correct behaviour for this rule.


To fix this in your project, you should define `validate` as a function property rather than a method. This means that it has no implicit this and won't be reported by the rule.

Thanks
